lib/transactions/transaction.ts

```diff
@@ -0,0 +1,228 @@
+    REJECTED = 6
+}
+
+/**
+ * Any transaction has 5 steps:
+ * 1. Create request JSON
+ * 2. Process request JSON to an unsigned transaction (remote API call to a node is necessary)
+ * 3. Sign the unsigned transaction
+ * 4. Broadcast the signed transaction (remote API call to a node is necessary)
+ * 5. [Optional] Transaction is confirmed after the trasaction is written to the blockchain and at leat
+ *    one block is written on top of the transaction block (remote API call to a node is necessary).
+ * 
+ * The state of the transaction can only go through each step in the specified order.
+ */
+export class Transaction {
+    private _requestJSON: IRequestJSON | null = null;
+    private _unsignedTransactionBytes: string | null = null;
+    private _signedTransactionBytes: string | null = null;
+    private _state: TransactionState;
+    private _transactionID: string | null = null;
+    private _fullHash: string | null = null;
+    private _transactionJSON: ITransactionJSON | null = null;
+
+
+    //========== Step 1 (local)=============
+    protected constructor(requestJSON?: IRequestJSON) {
+        if (requestJSON) {
+            this._requestJSON = requestJSON;
+            this._state = TransactionState.REQUEST_CREATED;
+        } else {
+            this._state = TransactionState.ERROR;
+        }
+    }
+
+    async calculateFee(remote: RemoteAPICaller) {
+        const data = await remote.apiCall('post', { ...this.requestJSON, feeNQT: '0' } as unknown as Record<string, string>);
+        const transactionData = data as unknown as ITransaction;
+        return CryptoUtil.Crypto.NqtToGmd(transactionData.transactionJSON.feeNQT);
+    }
+
+    setFee(feeGMD: string) {
+        if (this.state === TransactionState.REQUEST_CREATED) {
+            (this.requestJSON as IRequestJSON).feeNQT = CryptoUtil.Crypto.GmdToNqt(feeGMD);
+        } else {
+            throw new Error('Cannot set fee after the unsigned transaction was already created');//TODO refine errors
+        }
+    }
+    //======================================
+
+
+    //========== Step 2 (remote)=============
+    async createUnsignedTransaction(remote: RemoteAPICaller) {
+        if (this.canCreateUnsignedTransaction()) {
+            const unsignedTransaction = await remote.apiCall('post', this.requestJSON as unknown as Record<string, string>);
+            this.#onCreatedUnsignedTransaction((unsignedTransaction as unknown as IUnsignedTransaction).unsignedTransactionBytes);
+        } else {
+            throw new Error('createUnsignedTransaction cannot be processed. transaction=' + JSON.stringify(this));
+        }
+    }
+
+    canCreateUnsignedTransaction(): boolean {
+        if (this._requestJSON && 'secretPhrase' in this._requestJSON) {
+            throw new Error('Do not send secret password to node!');
+        }
+        return this.state === TransactionState.REQUEST_CREATED;
+    }
+
+    #onCreatedUnsignedTransaction(unsignedTransactionBytes: string) {

```

Comment:
is the hastag intentional?

Reply:
Yes. I used # for all private methods.

Reply:
Replaced # with "private" as there were some issues with the tests and version of ES supporting this # feature.